python/cudf/cudf/core/buffer.py

```diff
@@ -1,218 +1,61 @@
-            arg = utils.normalize_index(int(arg), self.size)
-            item = self.mem[arg]
-            if isinstance(item, str):
-                return item
-            # the dtype argument is necessary for datetime64 support
-            # because currently we can't pass datetime64 types into
-            # cuda dev arrays, so the type of the cuda dev array is
-            # an i64, and we view it as the dtype on the buffer
-            return item.view(self.dtype)
+    @classmethod
+    def from_array_like(cls, data):
+        if hasattr(data, "__cuda_array_interface__"):
+            ptr, size = _buffer_data_from_array_interface(
+                data.__cuda_array_interface__
+            )
+            return Buffer(ptr, size, owner=data)
+        elif hasattr(data, "__array_interface__"):
+            ptr, size = _buffer_data_from_array_interface(
+                data.__array_interface__
+            )
+            dbuf = DeviceBuffer(ptr, size)
+            return Buffer.from_device_buffer(dbuf)
         else:
-            raise NotImplementedError(type(arg))
-
-    @property
-    def avail_space(self):
-        return self.capacity - self.size
-
-    def _sentry_capacity(self, size_needed):
-        if size_needed > self.avail_space:
-            raise MemoryError("insufficient space in buffer")
-
-    def append(self, element):
-        self._sentry_capacity(1)
-        self.extend(np.asarray(element, dtype=self.dtype))
-
-    def extend(self, array):
-        from cudf.core.column import column
-
-        needed = array.size
-        self._sentry_capacity(needed)
-
-        array = column.as_column(array).astype(self.dtype).data.mem
-
-        self.mem[self.size : self.size + needed].copy_to_device(array)
-        self.size += needed
-
-    def to_array(self):
-        return self.mem[: self.size].copy_to_host()
-
-    def to_gpu_array(self):
-        return self.mem[: self.size]
-
-    def copy(self):
-        """Deep copy the buffer
-        """
-        return Buffer(
-            mem=cudautils.copy_array(self.mem),
-            size=self.size,
-            capacity=self.capacity,
-        )
-
-    def as_contiguous(self):
-        out = Buffer(
-            mem=cudautils.as_contiguous(self.mem),
-            size=self.size,
-            capacity=self.capacity,
-        )
-        assert out.is_contiguous()
-        return out
-
-    def is_contiguous(self):
-        return self.mem.is_c_contiguous()
-
-    def astype(self, dtype):
-        from cudf.core.column import column
-
-        return column.as_column(self).astype(dtype).data
-
-
-class BufferSentryError(ValueError):
-    pass
-
+            raise TypeError(
+                f"Cannot construct Buffer from {data.__class__.__name__}"
+            )
 
-class _BufferSentry(object):
-    def __init__(self, buf):
-        self._buf = buf
+    @classmethod
+    def from_device_buffer(cls, dbuf):
+        assert isinstance(dbuf, DeviceBuffer)
```

Comment:
Should this raise a `TypeError` if the type doesn't match?

Reply:
Yup

Reply:
We just use the `Buffer` constructor directly now